### ISSUE TITLE
activation_offloading: true causes linear VRAM leak with Gemma 4 26B-A4B (LoRA + QLoRA, with torch_compile + sample_packing)

### ISSUE THREAD

**Issue by grgkovac** [axolotl-ai-cloud/axolotl#3638]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

GPU VRAM usage should remain approximately constant across training steps after warmup. 

### Current behaviour

When fine-tuning `google/gemma-4-26B-A4B` with `activation_offloading: true`, GPU VRAM grows linearly throughout training until OOM. The leak reproduces with both LoRA and QLoRA adapters. 

This is **not** the per-checkpoint spike pattern reported in #3596 — no checkpoint saves or evaluation are configured in the runs below. 

The yaml is essentially https://github.com/axolotl-ai-cloud/axolotl/blob/main/examples/gemma4/26b-a4b-moe-qlora.yaml 
I am using cu130 because of B300 see #3630 

### Steps to reproduce

Six near-identical configs were run side-by-side, differing only in adapter type (LoRA vs QLoRA) and one toggled feature:

| Run | Adapter | `activation_offloading` | `use_scattermoe` | Result |
|---|---|---|---|---|
| `qlora-control` | qlora | true | true | **leaks** linearly (24 → 260 GiB over 293 steps) |
| `qlora-no-actoffload` | qlora | **false** | true | **flat at 24 GiB** |
| `qlora-no-scattermoe` | qlora | true | **false** | flat at 100 GiB |
| `lora-control` | lora | true | true | **leaks** linearly (120 → 260 GiB over 224 steps) |
| `lora-no-actoffload` | lora | **false** | true | **flat at 60 GiB** |
| `lora-no-scattermoe` | lora | true | **false** | **leaks** linearly (120 → 260 GiB over 149 steps) |

`activation_offloading: false` is the only intervention that stops the leak in both LoRA and QLoRA. Disabling ScatterMoE only stops the leak in the QLoRA case.

See attached TensorBoard screenshots showing `train/memory/max_allocated`, `train/memory/device_reserved`, `train/memory/max_active`, and `train/loss` for both matrices.

QLora
<img width="1568" height="766" alt="Image" src="https://github.com/user-attachments/assets/7f3801f0-ef75-4541-9ef7-7fa597c0770f" /> 
Lora
<img width="1568" height="766" alt="Image" src="https://github.com/user-attachments/assets/5747c09a-ce0f-43f4-84c0-ea0b79b22485" />

Dockerfile (cu130 was needed for  NVIDIA B300 SXM6 AC see issue #3630 )

```
FROM axolotlai/axolotl-uv:main-py3.11-cu130-2.9.1

RUN uv pip install --python /workspace/axolotl-venv/bin/python \
    "cut-cross-entropy[transformers] @ git+https://github.com/axolotl-ai-cloud/ml-cross-entropy.git@fec1a88"
```

Docker run
```
# launch train
docker run --rm --gpus "$GPU_SPEC" --ipc=host \
  --shm-size 10g \
  -e TRITON_PTXAS_PATH=/usr/local/cuda/bin/ptxas \
  -e HF_TOKEN="$HF_TOKEN" \
  -v "$HOME/projects/project-training:/workspace/project" \
  -v "$HOME/.cache/huggingface:/root/.cache/huggingface" \
  -w /workspace/project \
  project-axolotl \
  axolotl train "$CONFIG"
```

yamls
[bug_lora-noactoffload.yaml](https://github.com/user-attachments/files/27277714/bug_lora-noactoffload.yaml)
[bug_lora-noscattermoe.yaml](https://github.com/user-attachments/files/27277710/bug_lora-noscattermoe.yaml)
[bug_lora.yaml](https://github.com/user-attachments/files/27277711/bug_lora.yaml)
[bug_qlora-noactoffload.yaml](https://github.com/user-attachments/files/27277713/bug_qlora-noactoffload.yaml)
[bug_qlora-noscattermoe.yaml](https://github.com/user-attachments/files/27277709/bug_qlora-noscattermoe.yaml)
[bug_qlora.yaml](https://github.com/user-attachments/files/27277712/bug_qlora.yaml)

### Config yaml

```yaml
# BUG REPRO: Gemma 4 26B-A4B MoE QLoRA — CONTROL (expected: VRAM leak)

base_model: google/gemma-4-26B-A4B

plugins:
  - axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
  - axolotl.integrations.kernels.KernelsPlugin
  - axolotl.integrations.liger.LigerPlugin
use_kernels: true
use_scattermoe: true
experts_implementation: scattermoe
torch_compile: true
liger_layer_norm: true
liger_rope: true
liger_rms_norm: true
liger_glu_activation: true
liger_rms_norm_gated: true
strict: false

chat_template: gemma4
datasets:
  - path: mlabonne/FineTome-100k
    type: chat_template
    split: train[:10%]
    field_messages: conversations
    message_property_mappings:
      role: from
      content: value

val_set_size: 0
output_dir: ./outputs/bug/qlora-control

sequence_len: 2048
sample_packing: true

load_in_4bit: true
quantize_moe_experts: true
adapter: qlora
lora_r: 16
lora_alpha: 32
lora_dropout: 0
lora_target_modules: 'model.language_model.layers.[\d]+.(_checkpoint_wrapped_module.)?(mlp|self_attn).(up|down|gate|q|k|v|o)_proj'
lora_target_parameters:
  - experts.gate_up_proj
  - experts.down_proj
lora_mlp_kernel: false
lora_qkv_kernel: false
lora_o_kernel: false

bnb_config_kwargs:
  bnb_4bit_use_double_quant: true

gradient_accumulation_steps: 4
micro_batch_size: 1
num_epochs: 1
optimizer: adamw_torch_8bit
lr_scheduler: cosine
learning_rate: 0.0002

bf16: auto
tf32: true

gradient_checkpointing: true
activation_offloading: true

sdp_attention: true

warmup_ratio: 0.1
evals_per_epoch: 0
saves_per_epoch: 0
save_strategy: "no"
weight_decay: 0.0
special_tokens:

logging_steps: 1
use_tensorboard: true
run_name: qlora-control
```

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

main @ 6136ae6 (image: axolotlai/axolotl-uv:main-py3.11-cu130-2.9.1)

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by ved1beta** [comment]
thanks for reporting this , docker fix is kind of done #3641  . we also have a vram leak patch #3635 
these should be enough i assume you can have look.

**Comment by ved1beta** [comment]
this should be fixed please can you retry with LTS trl
